shellai/config.py

```diff
@@ -0,0 +1,150 @@
+import json
+import logging
+import sys
+from collections import namedtuple
+from pathlib import Path
+
+if sys.version_info >= (3, 11):
+    import tomllib
+else:
+    import tomli as tomllib
+
+CONFIG_DEFAULT_PATH: Path = Path("~/.config/shellai/config.toml")
+
+# tomllib does not support writting files, so we will create our own.
+CONFIG_TEMPLATE = """
+[output]
+# otherwise recording via script session will be enforced
+enforce_script = {enforce_script}
+# file with output(s) of regular commands (e.g. ls, echo, etc.)
+file = "{output_file}"
+# Keep non-empty if your file contains only output of commands (not prompt itself)
+prompt_separator = "{prompt_separator}"
+
+[history]
+enabled = {enabled}
+file = "{history_file}"
+# max number of queries in history (including responses)
+max_size = {max_size}
+
+[backend]
+endpoint = "{endpoint}"
+
+"""
+
+
+class Output(namedtuple("Output", ["enforce_script", "file", "prompt_separator"])):
+    """This class represents the [output] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        enforce_script: bool = False,
+        file: Path = Path("/tmp/shellai_output.txt"),
+        prompt_separator: str = "$",
+    ):
+        return super(Output, cls).__new__(cls, enforce_script, file, prompt_separator)
+
+
+class History(namedtuple("History", ["enabled", "file", "max_size"])):
+    """This class represents the [history] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        enabled: bool = True,
+        file: Path = Path("~/.local/share/shellai/shellai_history.json"),
+        max_size: int = 100,
+    ):
+        file = Path(file)
+        return super(History, cls).__new__(cls, enabled, file, max_size)
+
+
+class Backend(namedtuple("Backend", ["endpoint"])):
+    """This class represents the [backend] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
+        cls,
+        endpoint: str = "http://0.0.0.0:8080/v1/query/",
+    ):
+        return super(Backend, cls).__new__(cls, endpoint)
+
+
+class Config:
+    """Class that holds our configuration file representation.
+
+    .. note::
+        With this class, after being initialized, one can access their fields like:
+
+        >>> config = Config()
+        >>> config.output.enforce_script
+
+        The currently available top-level fields are:
+            * output = Match the `py:Output` class and their fields
+            * history = Match the `py:History` class and their fields
+            * backend = Match the `py:backend` class and their fields
+    """
+
+    def __init__(self, output: dict, history: dict, backend: dict) -> None:
+        self.output: Output = Output(**output)
+        self.history: History = History(**history)
+        self.backend: Backend = Backend(**backend)
```

Comment:
Note: There are two things to think about when coding the construction of `Config` in this way:

1. There's no way to tell whether a value was explicitly set by the user or was added by default values
2. There's no way for the user to unset a config value.

Neither of these are necessarily problems.  They are limitations of what you can do with the Config framework later on and could cause you to have to write some awkward code if you need to allow those things for a particular Config value later. (For instance, setting the default in the `NamedTuple` to `None` and then keeping the real default value for that specific config value elsewhere.)

Just thought you should be aware of those caveats in case you think they are useful features that you should build into the framework.